### PR TITLE
MVP: Testcase asset generation

### PR DESCRIPTION
**Pull request by woodruffw**
This builds up a little API + dependency injection scheme for declaring testcase assets.

~~The `_asset` decorator itself is pretty gnarly, but encapsulates a lot of niceties that'll be useful as we expand the test suite: it caches each test asset as it's lazily evaluated, handles parametric assets, etc.~~

See #3.

### CODE REVIEW

**Comment by woodruffw** [comment]
~~NB: I haven't checked in any generated assets yet, since I still need to figure out the best way to handle re-loading of assets that already exist on disk. This in turn is necessary because we probably don't want to blow away to whole asset suite each time a key changes.~~

**Comment by woodruffw** [comment]
I'll also use this PR to start the actual testcase development, since it's closely tied to the assets.

File: testcases/ee-certs-pathlen-constraints.yml
Comment (woodruffw) [inline testcases/ee-certs-pathlen-constraints.yml:1]: This isn't validated yet, just bringing up for review:

1. This is YAML at the moment, which is nobody's favorite. I did that just because editing JSON is more tedious + these individual testcase files will be merged into one giant JSON blob at the end, anyways. Do you have a preference for this being something else, e.g. TOML or just JSON?
2. What do you think about stable identifiers for individual testcases? I can imagine things becoming hard to follow once multiple individual testcase files are merged or used together; each testcase could carry a short unique `id` field to distinguish it.
3. More generally: what do you think about merging vs. not merging testcases? The original sketch in #3 includes merging `testcases/*` into one big `limbo.json` or similar file, but I'm wondering what advantages that presents versus a bunch of smaller (and therefore more editable) testcase files. I don't even fully remember where the merging idea came from; that might be purely a product of my own imagination 😅

File: limbo/_cli.py [hunk 1]
```diff
@@ -1,20 +1,124 @@
...
+    testcase_dir: Path = args.testcases.resolve()
+
+    output_path = testcase_dir / "limbo.json"
+    if output_path.exists() and not args.force:
+        _die(f"not overwriting {output_path} without --force")
+
+    # NOTE: Paths in testcases are relative to the testcase directory,
+    # so we chdir to pass validation.
+    all_testcases: list[Testcase] = []
+    with contextlib.chdir(testcase_dir):
```
Comment (alex) [inline limbo/_cli.py]: On principle I would say that you should basically never `chdir`

File: limbo/assets.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+from typing import Any, Callable, Iterable, ParamSpec, Self, Sequence, TypeVar, overload
+
+from cryptography import x509
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.hazmat.primitives.asymmetric.types import PrivateKeyTypes
+from cryptography.x509 import NameOID
+
+_EPOCH = datetime.datetime.fromtimestamp(0)
+_ONE_THOUSAND_YEARS_OF_TORMENT = _EPOCH + datetime.timedelta(days=365 * 1000)
```
Comment (alex) [inline limbo/assets.py:19]: this isn't actually 1000 years precisely because of leap days.

thank you for coming to my ted talk

File: limbo/models.py [hunk 3]
```diff
@@ -2,15 +2,20 @@
...
 from datetime import datetime
 from enum import Enum
+from pathlib import Path
 from typing import Literal
 
-from pydantic import BaseModel, ConstrainedStr, Field, StrictInt, StrictStr
+from pydantic import BaseModel, ConstrainedStr, Field, StrictStr, validator
 
-PeerKind = Literal["RFC822"] | Literal["DNS"] | Literal["IP"]
+PeerKind = Literal["RFC822"] | Literal["DNS"] | Literal["IP"] | Literal["DN"]
```
Comment (alex) [inline limbo/models.py]: Is DN here DirectoryName? I'm not sure I've ever seen one in reality, so I'm skeptical they need to be part of this initial spec.

File: testcases/ee-certs-pathlen-constraints.limbo.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
+version: 1
+testcases:
+  - id: ee-pathlen-0
```
Comment (alex) [inline testcases/ee-certs-pathlen-constraints.limbo.yml]: So, these feel pretty verbose, and very far from the definition of the assets to me.

My inclination is that the assets and the test cases should be much closer together, since they are deeply inter-related.

While I do not recommend following this exactly, I think the example in https://github.com/alex/x509-validator/blob/master/tests/test_basic.py is closer to what I'd expect.

File: limbo/models.py (hunk 3 above)
Comment (woodruffw) [reply limbo/models.py]: That was meant to be "distinguished name," i.e. a subject instead of a SAN in case the peer doesn't contain a SAN. But maybe that's too flexible here/should be another field?

File: testcases/ee-certs-pathlen-constraints.limbo.yml (hunk 4 above)
Comment (woodruffw) [reply testcases/ee-certs-pathlen-constraints.limbo.yml]: Yeah, I was thinking about that -- one thing I could do is have `limbo compile` mean "embed the assets into their testcases". This would reduce the amount of testcase state down to a single unified JSON file, at the cost of some bloat in that file.

Thoughts?

File: limbo/models.py (hunk 3 above)
Comment (alex) [reply limbo/models.py]: Subjects shouldn't be used for this purpose at all at this point.

File: testcases/ee-certs-pathlen-constraints.limbo.yml (hunk 4 above)
Comment (alex) [reply testcases/ee-certs-pathlen-constraints.limbo.yml]: I'd go the opposite direction: Put the test case generation in Python, next to the asset generation. WDYT?

File: limbo/_cli.py (hunk 1 above)
Comment (woodruffw) [reply limbo/_cli.py]: Agreed -- this was a hack, but I wasn't too sure that it was worth improving (especially w/ bundling like suggested in https://github.com/trailofbits/x509-limbo/pull/5#discussion_r1206038345).

But if you'd like to ban it on principle, I can refactor.

Comment (alex) [reply limbo/_cli.py]: It seems like its entirely a matter of path handling, so yes, I think it'd be better to handle properly.

File: testcases/ee-certs-pathlen-constraints.limbo.yml (hunk 4 above)
Comment (woodruffw) [reply testcases/ee-certs-pathlen-constraints.limbo.yml]: That sounds good to me! Are you thinking it'd mirror the current asset generation, i.e. each would essentially distill into a `Testcase` object that could then be serialized?

Comment (alex) [reply testcases/ee-certs-pathlen-constraints.limbo.yml]: Yup!

Comment (woodruffw) [reply testcases/ee-certs-pathlen-constraints.limbo.yml]: Cool, that sounds a lot nicer.

Just to sketch a bit, something like this?

```python
@testcase("some-id")
def whatever(workspace):
    workspace.trusted_certs([some_ca(), another_ca()])
    workspace.untrusted_intermediates(...)
    
    # etc.
    
    return workspace.testcase()
```

Comment (alex) [reply testcases/ee-certs-pathlen-constraints.limbo.yml]: I think that _general_ design works. I don't want to be overly perscriptive in terms of a workspace vs. some other API design, but I think that design of Python function that buidls a test case based invoking the assets is A+

Comment (woodruffw) [reply testcases/ee-certs-pathlen-constraints.limbo.yml]: Yeah, `workspace` was just me thinking about phases between an intermediate testcase and a "reified" one. I can probably make it nicer than that.

Either way, I can definitely start refactoring towards that!

File: limbo/assets.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    """
+
+    cert: x509.Certificate
+    key: PrivateKeyTypes
+
+    @cached_property
+    def key_pem(self) -> str:
+        return self.key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
```
Comment (alex) [inline limbo/assets.py:38]: Future us problem, but this wont' work if we ever want to do ed25519 keys.

File: limbo/testcases/pathlen.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    builder = builder.client_validation()
+    builder = (
+        builder.trusted_certs(root)
+        .untrusted_intermediates(intermediate)
+        .peer_certificate(leaf)
+        .succeeds()
+    )
+
+    builder.succeeds()
```
Comment (alex) [inline limbo/testcases/pathlen.py]: Double succeeds?

**Review by alex — COMMENTED** [review]
Conceptually this all is looking good to me

File: limbo/testcases/pathlen.py (hunk 6 above)
Comment (woodruffw) [reply limbo/testcases/pathlen.py]: Yep, refactoring error. Fixed!

File: limbo/assets.py (hunk 5 above)
Comment (woodruffw) [reply limbo/assets.py:38]: Yep -- I need to make this more generic, but that's a non-MVP problem 🙂
